src/rpc/blockchain.cpp

```diff
@@ -2290,6 +2290,198 @@ static RPCHelpMan scantxoutset()
+        GCSFilter::ElementSet needle_set;
+        for (const UniValue& scanobject : request.params[1].get_array().getValues()) {
+            FlatSigningProvider provider;
+            std::vector<CScript> scripts = EvalDescriptorStringOrObject(scanobject, provider);
+            for (const CScript& script : scripts) {
+                needle_set.emplace(script.begin(), script.end());
+            }
+        }
+        NodeContext& node = EnsureNodeContext(request.context);
+        UniValue blocks(UniValue::VARR);
+        const int amount_per_chunk = 10000;
+        const CBlockIndex* start_index = block; // for remembering the start of a blockfilter range
+        std::vector<BlockFilter> filters;
+        const CBlockIndex* start_block = block; // for progress reporting
+        const CBlockIndex* last_scanned_block = block;
+        g_scanfilter_should_abort_scan = false;
+        g_scanfilter_progress = 0;
+        g_scanfilter_progress_height = start_block->nHeight;
+        while (block) {
+            node.rpc_interruption_point(); // allow a clean shutdown
+            if (g_scanfilter_should_abort_scan) {
+                break;
+            }
+            const CBlockIndex* next = nullptr;
+            {
+                LOCK(cs_main);
+                next = ChainActive().Next(block);
+                if (block == stop_block) next = nullptr;
+            }
+            if (start_index->nHeight+amount_per_chunk == block->nHeight || next == nullptr) {
+                LogPrint(BCLog::RPC, "Fetching blockfilters from height %d to height %d.\n", start_index->nHeight, block->nHeight);
+                if (index->LookupFilterRange(start_index->nHeight, block, filters)) {
+                    for (const BlockFilter& filter : filters) {
+                        // compare the elements-set with each filter
+                        if (filter.GetFilter().MatchAny(needle_set)) {
+                            blocks.push_back(filter.GetBlockHash().GetHex());
+                            LogPrint(BCLog::RPC, "scanblocks: found match in %s\n", filter.GetBlockHash().GetHex());
+                        }
+                    }
+                }
+                start_index = block;
+
+                // update progress
+                int blocks_processed = block->nHeight - start_block->nHeight;
+                int total_blocks_to_process = stop_block->nHeight - start_block->nHeight;
+                if (total_blocks_to_process > 0) { // avoid division by zero
+                    g_scanfilter_progress = (int)(100.0 / total_blocks_to_process * blocks_processed);
+                } else {
+                    g_scanfilter_progress = 100;
+                }
+                g_scanfilter_progress_height = block->nHeight;
+            }
+            last_scanned_block = block;
+            block = next;
+        }
+        ret.pushKV("from_height", start_block->nHeight);
+        ret.pushKV("to_height", last_scanned_block->nHeight);
```

Comment:
Not sure about this. The interruption point is at the beginning of the while loop. Processing blocks (scanning filters) can't be done without setting `last_scanned_block = block;`. Shouldn't therefore `last_scanned_block` always point to `stopindex` of `LookupFilterRange()` (a.k.a. the last scanned block)? Or do I miss something?

Reply:
Sorry, that I didn't get around to answer earlier. I am trying to simplify so please ignore if I get some minor details wrong: Let's say we are scanning blocks 10.000 - 30.000. The while loop runs through every block and with each block `last_scanned_block` is updated. But no scanning is actually happening. At 20.000 it does the actual scanning for the first time, blocks 10k-20k. So at height 20k `last_scanned_block` is actually in sync with that, up to 20k was scanned and the `last_scanned_block` is 20k. From the next block on `last_scanned_block` increments with every loop again and is out of sync with what was actually synced. So if the interruption point is hit at, for example, 25k, then `to_height` would say 25k but actually, the scan was only done up to 20k. One very simple fix would be to move `last_scanned_block = block;
` just 2 lines higher so it is only updated after an actual scan. But then it is just reports the same as `g_scanfilter_progress_height` so I think it can be dropped altogether. I don't see `last_scanned_block` being used anywhere aside for the `to_height` return value usage in this line here, so I am not sure what you mean by "it can't be done without it". And just fundamentally I don't see why this would not be the same: The progress height and the last scanned block are the same value to me.

I just realized I am writing a whole paragraph for a 3 line change, code is probably much easier to reason about :) https://github.com/fjahr/bitcoin/commit/16c7cb5b67e95ba1c8050eacfd4099fad108acf3 Tests pass.

Reply:
In 6a69dd267e4b96e6657de7b77521311994b2a902 "Add scanblocks RPC call - scan for relevant blocks with descriptors"

I agree with @fjahr here. `last_scanned_block` isn't actually the last block we have scanned, and if we abort, then `to_height` does not accurately reflect where we have actually scanned up to. It would be better to move it up into the `if` the guards the actual scanning. But in that case, it's the same as `g_scanfilter_progress_height`, so it can just be removed.

Reply:
Fixed in #23549